### PR TITLE
Instruction cache pollution for distbench 

### PR DESCRIPTION
**Pull request by AbhayPatil**
Instruction cache pollution for distbench.

### CODE REVIEW

File: activity.h [hunk 1]
```diff
@@ -94,9 +94,35 @@ struct PolluteDataCacheConfig {
...
+  int array_reads_per_iteration_ = 0;
+  int max_func_num_ = 0;
+  std::mt19937 mersenne_twister_prng_;
+  std::vector<MyFunctionPtr> func_ptr_array_;
+  int64_t optimization_preventer_ = 0;
+  std::uniform_int_distribution<> random_index_;
+};
+
+struct PolluteInstructionCacheConfig {
+  int array_reads_per_iteration;
```
Comment (danmanj) [inline activity.h:118]: array_reads is a poort name for this variable. Functions invoked maybe?

File: activity.cc [hunk 2]
```diff
@@ -129,4 +132,99 @@ ActivityLog PolluteDataCache::GetActivityLog() {
...
+      func_ptr_array_.size();
+  random_index_ = std::uniform_int_distribution<>(0, max_func_num_ - 1);
+
+  iteration_count_ = 0;
+}
+
+void PolluteInstructionCache::DoActivity() {
+  iteration_count_++;
+  int64_t sum = 0;
+
```
Comment (danmanj) [inline activity.cc:211]: delete extra blank line

File: activity.cc [hunk 3]
```diff
@@ -129,4 +132,99 @@ ActivityLog PolluteDataCache::GetActivityLog() {
   return alog;
 }
 
+// Activity: PolluteInstructionCache
+
+// The boost library (https://github.com/nelhage/rules_boost) is used to
+// create a large 2D array of functions with names -
```
Comment (danmanj) [inline activity.cc:138]: how large?

File: activity.cc [hunk 4]
```diff
@@ -129,4 +132,99 @@ ActivityLog PolluteDataCache::GetActivityLog() {
...
+// ...
+// ...
+// ...
+// DummyFunc_{POLLUTE_ICACHE_LOOP_SIZE-1}_0, ..., DummyFunc_{POLLUTE_ICACHE_LOOP_SIZE-1}_{POLLUTE_ICACHE_LOOP_SIZE-1}.
+// Each of these functions returns an int.
+//
+// This 2D array was created because BOOST_PP_REPEAT() macro can
+// generate at most 1024 repetitions.
+//
+// The purpose of 'do_work' part of the function is only to increase
```
Comment (danmanj) [inline activity.cc:150]: this comment belongs next to the relevant code.

File: activity.cc [hunk 5]
```diff
@@ -129,4 +132,99 @@ ActivityLog PolluteDataCache::GetActivityLog() {
...
+// create a large 2D array of functions with names -
+//
+// DummyFunc_0_0, DummyFunc_0_1, ..., DummyFunc_0_{POLLUTE_ICACHE_LOOP_SIZE-1},
+// ...
+// ...
+// ...
+// DummyFunc_{POLLUTE_ICACHE_LOOP_SIZE-1}_0, ..., DummyFunc_{POLLUTE_ICACHE_LOOP_SIZE-1}_{POLLUTE_ICACHE_LOOP_SIZE-1}.
+// Each of these functions returns an int.
+//
+// This 2D array was created because BOOST_PP_REPEAT() macro can
```
Comment (danmanj) [inline activity.cc:147]: "foo was created" is passive voice. "we create foo" is better

File: activity.cc [hunk 6]
```diff
@@ -129,4 +132,99 @@ ActivityLog PolluteDataCache::GetActivityLog() {
...
+    } else                        \
+      return n;                   \
+  }
+#define GENERATE_FUNC_OUTER_LOOP(z, n, text) \
+  BOOST_PP_REPEAT(POLLUTE_ICACHE_LOOP_SIZE, GENERATE_FUNC_INNER_LOOP, n)
+
+// Get the pointers to the functions created above and store it in
+// func_ptr_array.
+#define GET_FUNC_PTR_INNER_LOOP(z, n, prefix) \
+  func_ptr_array_[(prefix * POLLUTE_ICACHE_LOOP_SIZE + n)] = &DummyFunc_ ## prefix ## _ ## n;
```
Comment (danmanj) [inline activity.cc:177]: use the \ at the end of line to make the macro fit in 80 columns

File: activity.h (hunk 1 above)
Comment (AbhayPatil) [reply activity.h:118]: changed it to - function_invocations_per_iteration

File: activity.cc (hunk 2 above)
Comment (AbhayPatil) [reply activity.cc:211]: Done

File: activity.cc (hunk 5 above)
Comment (AbhayPatil) [reply activity.cc:147]: Done

File: activity.cc (hunk 6 above)
Comment (AbhayPatil) [reply activity.cc:177]: Done

File: activity.cc (hunk 3 above)
Comment (AbhayPatil) [reply activity.cc:138]: Changed language

**Comment by AbhayPatil** [comment]
After changing number of functions from 62k to 10k... the binary size has reduced from 19MB to 12MB

File: distbench_engine.cc [hunk 7]
```diff
@@ -179,6 +179,17 @@ absl::Status DistBenchEngine::ParseActivityConfig(ActivityConfig& ac) {
...
+    auto status = PolluteInstructionCache::ValidateConfig(ac);
+    if (!status.ok()) return status;
+
+    s.pollute_instruction_cache_config.function_invocations_per_iteration =
+        GetNamedSettingInt64(ac.activity_settings(),
+                             "function_invocations_per_iteration", 1000);
+
+    s.pollute_instruction_cache_config.max_func_num =
+        GetNamedSettingInt64(ac.activity_settings(),
+                             "max_func_num", 0);
```
Comment (danmanj) [inline distbench_engine.cc:192]: this name should not be abbreviated when the other isn't

Comment (AbhayPatil) [reply distbench_engine.cc]: Removed max_func_num setting for icache pollution.

**Review by danmanj — APPROVED** [review]

**Pull request merged**